pkg/services/screenshot/screenshot.go

```diff
@@ -0,0 +1,379 @@
+	Path string
+	URL  string
+}
+
+// ScreenshotOptions are the options for taking a screenshot.
+type ScreenshotOptions struct {
+	DashboardUID string
+	PanelID      int64
+	Width        int
+	Height       int
+	Theme        models.Theme
+	Timeout      time.Duration
+}
+
+// SetDefaults sets default values for missing or invalid options.
+func (s ScreenshotOptions) SetDefaults() ScreenshotOptions {
+	if s.Width <= 0 {
+		s.Width = DefaultWidth
+	}
+	if s.Height <= 0 {
+		s.Height = DefaultHeight
+	}
+	switch s.Theme {
+	case models.ThemeDark, models.ThemeLight:
+	default:
+		s.Theme = DefaultTheme
+	}
+	if s.Timeout <= 0 {
+		s.Timeout = DefaultTimeout
+	}
+	return s
+}
+
+// ScreenshotService is an interface for taking screenshots.
+//go:generate mockgen -destination=mock.go -package=screenshot github.com/grafana/grafana/pkg/services/screenshot ScreenshotService
+type ScreenshotService interface {
+	Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error)
+}
+
+// BrowserScreenshotService takes screenshots using a headless browser.
+type BrowserScreenshotService struct {
+	ds                  dashboards.DashboardService
+	rs                  rendering.Service
+	screenshotDuration  prometheus.Histogram
+	screenshotFailures  prometheus.Counter
+	screenshotSuccesses prometheus.Counter
+}
+
+func NewBrowserScreenshotService(metrics prometheus.Registerer, ds dashboards.DashboardService, rs rendering.Service) ScreenshotService {
+	return &BrowserScreenshotService{
+		ds: ds,
+		rs: rs,
+		screenshotDuration: promauto.With(metrics).NewHistogram(prometheus.HistogramOpts{
+			Name:      "browser_screenshot_duration_seconds",
+			Buckets:   []float64{0.1, 0.25, 0.5, 1, 2, 5, 10, 15},
+			Namespace: namespace,
+			Subsystem: subsystem,
+		}),
+		screenshotFailures: promauto.With(metrics).NewCounter(prometheus.CounterOpts{
+			Name:      "browser_screenshot_failures_total",
+			Namespace: namespace,
+			Subsystem: subsystem,
+		}),
+		screenshotSuccesses: promauto.With(metrics).NewCounter(prometheus.CounterOpts{
+			Name:      "browser_screenshot_successes_total",
+			Namespace: namespace,
+			Subsystem: subsystem,
+		}),
+	}
+}
+
+// Take returns a screenshot or an error if either the dashboard does not exist
+// or it failed to screenshot the dashboard. It uses both the context and the
+// timeout in ScreenshotOptions, however the timeout in ScreenshotOptions is
+// sent to the remote browser where it is used as a client timeout.
+func (s *BrowserScreenshotService) Take(ctx context.Context, opts ScreenshotOptions) (*Screenshot, error) {
+	q := models.GetDashboardQuery{Uid: opts.DashboardUID}
+	if err := s.ds.GetDashboard(ctx, &q); err != nil {
+		defer s.screenshotFailures.Inc()
+		return nil, err
+	}
+
+	start := time.Now()
+	defer func() { s.screenshotDuration.Observe(time.Since(start).Seconds()) }()
+
+	opts = opts.SetDefaults()
+
+	// Compute the URL to screenshot.
+	renderPath := path.Join("d-solo", q.Result.Uid, q.Result.Slug)
```

Comment:
Nit: Is there a constant we can re-use somewhere for this URL segment?

Reply:
Sure isn't. This is a broad problem in our entire API setup of the code - there are almost no path constants or programmatic construction anywhere. I don't know where I'd put the constants right now, tbh.